core/lib/zk_os_merkle_tree/src/hasher/proofs.rs

```diff
@@ -0,0 +1,393 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    iter,
+};
+
+use anyhow::Context;
+use zksync_basic_types::H256;
+
+use crate::{types::Leaf, BatchOutput, HashTree, TreeEntry};
+
+/// Operation on a Merkle tree entry used in [`BatchTreeProof`].
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum TreeOperation {
+    /// Update of an existing entry.
+    Update { index: u64 },
+    /// Insertion of a new entry.
+    Insert {
+        /// Prev index before *batch* insertion (i.e., always points to an index existing before batch insertion).
+        prev_index: u64,
+    },
+}
+
+#[derive(Debug)]
+struct InsertedRange {
+    old_next_index: u64,
+    keys_and_indices: BTreeMap<H256, u64>,
+}
+
+impl InsertedRange {
+    fn new(old_next_index: u64) -> Self {
+        Self {
+            old_next_index,
+            keys_and_indices: BTreeMap::new(),
+        }
+    }
+
+    fn update(&mut self, key: H256, index: u64) {
+        self.keys_and_indices.insert(key, index);
+    }
+}
+
+/// Merkle proof of batch insertion into [`MerkleTree`](crate::MerkleTree).
+///
+/// # How it's verified
+///
+/// Assumes that the tree before insertion is correctly constructed (in particular, leaves are correctly linked via prev / next index).
+/// Given that, proof verification is as follows:
+///
+/// 1. Check that all necessary leaves are present in `sorted_leaves`, and their keys match inserted / updated entries.
+/// 2. Previous root hash of the tree is recreated using `sorted_leaves` and `hashes`.
+/// 3. `sorted_leaves` are updated / extended as per inserted / updated entries.
+/// 4. New root hash of the tree is recreated using updated `sorted_leaves` and (the same) `hashes`.
+#[derive(Debug)]
+pub struct BatchTreeProof {
+    /// Performed tree operations. Correspond 1-to-1 to [`TreeEntry`]s.
+    pub operations: Vec<TreeOperation>,
+    /// Sorted leaves from the tree before insertion sufficient to prove it. Contains all updated leaves
+    /// (incl. prev / next neighbors for the inserted leaves), and the last leaf in the tree if there are inserts.
+    pub sorted_leaves: BTreeMap<u64, Leaf>,
+    /// Hashes necessary and sufficient to restore previous and updated root hashes. Provided in the ascending `(depth, index_on_level)` order,
+    /// where `depth == 0` are leaves, `depth == 1` are nodes aggregating leaf pairs etc.
+    pub hashes: Vec<H256>,
+}
+
+impl BatchTreeProof {
+    #[cfg(test)]
+    fn empty() -> Self {
+        Self {
+            operations: vec![],
+            sorted_leaves: BTreeMap::new(),
+            hashes: vec![],
+        }
+    }
+
+    /// Returns the new root hash of the tree on success.
+    pub fn verify(
```

Comment:
AFAIU, the verification procedure implemented in the ZK OS repo uses separate Merkle paths for each of `sorted_leaves`, so the proof there is longer, but IIUC, the amount of hashing is the same (e.g., Merkle paths are not folded separately for each leaf; they are used almost in the same way as `hashes` here).